h2o-core/src/main/java/hex/Model.java

```diff
@@ -2223,24 +2224,56 @@ public boolean testJavaScoring(Frame data, Frame model_predictions, EasyPredictM
     }
   }
 
+  static <T extends Lockable<T>> int deleteAll(Key<T>[] keys) {
+    int c = 0;
+    for (Key k : keys) {
+      T t = DKV.getGet(k);
+      if (t != null) {
+        t.delete(); //delete all subparts
+        c++;
+      }
+    }
+    return c;
+  }
+
   public void deleteCrossValidationModels() {
     if (_output._cross_validation_models != null) {
-      for (Key k : _output._cross_validation_models) {
-        Model m = DKV.getGet(k);
-        if (m!=null) m.delete(); //delete all subparts
+      Log.info("Cleaning up CV Models for " + this._key.toString());
```

Comment:
@seb-h2o it is okay to keep the keys in the object, it could be even argued that is the correct behavior - if someone reloads the object to memory you want to preserve the link to the model.

My previous comment was about not doing ` _output._cross_validation_models = null;` not about updating the model.

I think updating the model in this function is a bad idea, you cannot be sure what are the intentions of the caller and it the model should be updated (there can be other pending changes on the model) not just the change you are making in this functions.

It is perfectly fine to keep the references in the model. You cannot guarantee the consistency anyway - the user always can delete the CV models manually and "break" the reference.

Reply:
let's have a quick chat about this so that I can close the ticket

Reply:
made it more explicit here to avoid misuse

Reply:
TL;DR: agree with 3 - but with suggestions ;)

ad 1. Good points, in my opinion not a real issue. As I said previously arguments can be made that it actually does make sense to keep the keys in the model (to be eg. link them to to outputs in logs).
ad 2. I don't think it is "fiddling", models can be easily deleted even in Flow, the dead references are to be expected - what I am trying to say is that this is not an edge case.

And just to re-iterate my point of view and to clarify in general:
1. User builds a model, the model has CV models attached.
2. User decides to delete the CV models (by calling a function - we have to consider the Java API users!). IMHO modifying the model itself is not a good thing to do. We don't know what the user did with the models. They could have been saved for later or anything really. We should not destroy the link to the CV models or modify the model at all.

ad 3) Yes, we can handle it in AutoML. But I think there might be a better solution. Can we use flag  the `keep_cross_validation_models` in AutoML and set it to the model instances as well? The model can do the cleaning right after the main model is built when the Model object is still live and not finalized? We are trying to clean the model but we can prevent it from being "messy" in the first place.

Reply:
If you remove the problematic changes I think we can merge the PR, I don't see an issue with the rest.

But I am really not convinced the "nullyfying" is the right thing to do. If the model was trained with CV enabled,   I would expect an evidence that it actually happened that way. The CV models don't need to exist but I would expect keys to n CV models. That is no an unreasonable invariant. If you do not want to keep the keys, tell the model to clean the CV models in the process of model building. That also makes sense.

Reply:
this is why a quick chat is often more efficient than github :)
> Can we use flag the keep_cross_validation_models in AutoML and set it to the model instances as well?...` and the rest of this sentence.

This is exactly what I've been adding to `Model` and `ModelBuilder`. And this leads to CV models being consistently null in any case (automl or not) if the keep_cv_models param is set to false, and it behaves exactly how you suggeset in your last comment.
- Now comes the inconsistency in AutoML: the issue is not with `keep_cross_validation_models` but with `keep_cross_validation_predictions`. I believe those should behave the same way, and now they do in most cases. Except that AutoML requires to temporarily set `keep_cross_validation_predictions=true` to all models so that we can train the SE Models, and we can only clean up the predictions after the SE models are ready: this is where the `deleteCrossValidationPreds` enters the scene.

TBH, `deleteCrossValidationModels` is probably useless now, was a safety net and I believe it can be safely removed. But this is not the case for `deleteCrossValidationPreds`.

> If the model was trained with CV enabled, I would expect an evidence that it actually happened that way. The CV models don't need to exist but I would expect to see keys to the n CV models. That is not an unreasonable invariant.

I agree, this is not unreasonable to show some clue to the user that the models were computed using CV, but he has one: it's named `cross_validation_metrics`, those are always available and far more informative than dead keys ;)

To sum up:
1. I believe I can remove `deleteCrossValidationModels` completely as this one behaves consistently with this PR, just need to run a couple of tests and confirm it's safe to remove.
2. I'm not happy with `cross_validation_predictions` that behaves differently when training a single model or a grid, or AutoML. And also, it doesn't behave like `cross_validation_models`: I got convinced to "nullify" those predictions once I printed it with python to see `[ , , , , ]`. As a user, if I consume this, I don't know what to think about this empty-but-actually-not-really-empty array...

Reply:
Okay, can you then please undo the changes in `deleteCrossValidationModels`. Remove methods:

- deleteAll
- deleteCrossValidationModels(boolean nullify)

and revert all the changes in `deleteCrossValidationPreds`?

Happy to quick chat anytime - please ping me on Slack. Github comments are too asynchronous I usually go over them when I have free time.

Reply:
removed nullify/update and closing this